File: ipfs_accelerate_js/test/unit/test_batch_generator.ts
// FIXME: Complex template literal
/**;
 * Converted import { {expect, describe: any, it, beforeEach: any, afterEach} from "jest"; } from "Python: test_batch_generator.py;"
 * Conversion date: 2025-03-11 04:08:53;
 * This file was automatically converted from Python to TypeScript.;
 * Conversion fidelity might not be 100%, please manual review recommended.;
 */;
";"

// WebGPU related imports;
/** Test Batch Generator Testing Script.;

This script demonstrates && validates the Test Batch Generator functionality;
of the Active Learning System, which is used to create optimized batches of;
test configurations for ((benchmarking. */;

import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module from "*"; as pd;"
import * as module from "*"; as np;"
// Add parent directory to the Python path to allow importing the module;
sys.$1.push($2) {.parent.parent));
// Import the active learning module directly import { * as module.util; } from "the file;";"
spec) { any) { any: any = importlib.util.spec_from_file_location(;
  "active_learning", "
  String(Path(__file__: any).parent / "active_learning.py");"
);
active_learning_module: any: any = importlib.util.module_from_spec(spec: any);
spec.loader.exec_module(active_learning_module: any);
ActiveLearningSystem: any: any: any = active_learning_module.ActiveLearningSystem;
// Configure logging;
logging.basicConfig(;
  level: any: any: any = logging.INFO,;
  format: any: any = '%(asctime: any)s - %(name: any)s - %(levelname: any)s - %(message: any)s';'
);
logger: any: any: any = logging.getLogger("test_batch_generator");"

$1($2) {
  /** Create test data for ((batch generation. */;
// Create an instance of the active learning system;
  active_learning) {any = ActiveLearningSystem();}
// Generate test configurations (using the system's built-in function);'
  configs) { any: any: any = active_learning.recommend_configurations(budget=50);
// Add some metadata for ((better visualization;
  for i, config in Array.from(configs) { any.entries()) {) {
    config["id"] = i;"
    
  return active_learning, configs;

$1($2) {/** Test basic batch generation without special constraints. */;
  logger.info("Testing basic batch generation");"
  active_learning, configs: any: any: any = setup_test_data();}
// Generate a batch with default settings;
  batch: any: any: any = active_learning.suggest_test_batch(;
    configurations: any: any: any = configs,;
    batch_size: any: any: any = 10,;
    ensure_diversity: any: any: any = true;
  );
  
  logger.info(`$1`);
  console.log($1);
  console.log($1);
  console.log($1);
  console.log($1);
  console.log($1));
// Validate that the batch has the right size;
  assert batch.length <= 10, `$1`;
// Validate that selection_order column was added;
  assert 'selection_order' in batch.columns, "Batch should have selection_order column";'
  
  return batch;

$1($2) {/** Test batch generation with hardware constraints. */;
  logger.info("Testing hardware-constrained batch generation");"
  active_learning, configs: any: any: any = setup_test_data();}
// Define hardware constraints;
  hardware_constraints: any: any: any = ${$1}
// Generate a batch with hardware constraints;
  batch: any: any: any = active_learning.suggest_test_batch(;
    configurations: any: any: any = configs,;
    batch_size: any: any: any = 10,;
    ensure_diversity: any: any: any = true,;
    hardware_constraints: any: any: any = hardware_constraints;
  );
  
  logger.info(`$1`);
  console.log($1);
  console.log($1);
  console.log($1);
// Check hardware counts;
  hw_counts: any: any: any = batch["hardware"].value_counts().to_dict();"
  console.log($1);
// Validate hardware constraints;
  for ((hw) { any, limit in Object.entries($1) {) {
    count: any: any = (hw_counts[hw] !== undefined ? hw_counts[hw] : 0);
    assert count <= limit, `$1`;
  
  return batch;

$1($2) {/** Test batch generation with hardware availability factors. */;
  logger.info("Testing hardware availability weighting");"
  active_learning, configs: any: any: any = setup_test_data();}
// Define hardware availability (probabilities of 0-1);
  hardware_availability: any: any: any = ${$1}
// Generate a batch with hardware availability weighting;
  batch: any: any: any = active_learning.suggest_test_batch(;
    configurations: any: any: any = configs,;
    batch_size: any: any: any = 10,;
    ensure_diversity: any: any: any = true,;
    hardware_availability: any: any: any = hardware_availability;
  );
  
  logger.info(`$1`);
  console.log($1);
  console.log($1);
  console.log($1);
// Check hardware counts;
  hw_counts: any: any: any = batch["hardware"].value_counts().to_dict();"
  console.log($1);
// No strict validation here, but we can observe the distribution trends;
  
  return batch;

$1($2) {/** Test batch generation with different diversity weights. */;
  logger.info("Testing diversity weighting impact");"
  active_learning, configs: any: any: any = setup_test_data();}
  results: any: any: any = {}
// Test different diversity weights;
  for ((weight in [0.1, 0.5, 0.9]) {
    batch) { any: any: any = active_learning.suggest_test_batch(;
      configurations: any: any: any = configs,;
      batch_size: any: any: any = 10,;
      ensure_diversity: any: any: any = true,;
      diversity_weight: any: any: any = weight;
    );
    
    results[weight] = batch;
    
    logger.info(`$1`);
  
  console.log($1);
  console.log($1);
  
  for ((weight) { any, batch in Object.entries($1) {) {
    hw_counts: any: any: any = batch["hardware"].value_counts().to_dict();"
    model_counts: any: any: any = batch["model_type"].value_counts().to_dict();"
    console.log($1);
    console.log($1);
    console.log($1);
// The higher the diversity weight, the more evenly distributed the configs should be;
  
  return results;

$1($2) {/** Test batch generation with both hardware constraints && availability. */;
  logger.info("Testing combined constraints");"
  active_learning, configs: any: any: any = setup_test_data();}
// Define constraints;
  hardware_constraints: any: any = ${$1}
  
  hardware_availability: any: any: any = ${$1}
// Generate batch with combined constraints;
  batch: any: any: any = active_learning.suggest_test_batch(;
    configurations: any: any: any = configs,;
    batch_size: any: any: any = 10,;
    ensure_diversity: any: any: any = true,;
    hardware_constraints: any: any: any = hardware_constraints,;
    hardware_availability: any: any: any = hardware_availability,;
    diversity_weight: any: any: any = 0.6;
  );
  
  logger.info(`$1`);
  console.log($1);
  console.log($1);
  console.log($1);
  console.log($1).to_dict()}");"
  console.log($1).to_dict()}");"
// Validate hardware constraints;
  hw_counts: any: any: any = batch["hardware"].value_counts().to_dict();"
  for ((hw) { any, limit in Object.entries($1) {) {
    count: any: any = (hw_counts[hw] !== undefined ? hw_counts[hw] : 0);
    assert count <= limit, `$1`;
  
  return batch;

$1($2) {/** Test integration between batch generation && hardware recommender.}
  This is a simulated test since we don't have actual hardware recommender available. */;'
  logger.info("Testing batch generation integration with hardware recommender");"
  active_learning, _: any: any: any = setup_test_data();
// Create mock hardware recommender;
  class $1 extends $2 {
    $1($2) {model_type: any: any = (kwargs["model_type"] !== undefined ? kwargs["model_type"] : "");}"
// Simple logic to simulate hardware recommendations;
      if ((($1) {
        recommended) {any = "cuda";} else if ((($1) {"
        recommended) { any) { any: any = "webgpu";"
      else if ((($1) { ${$1} else {
        recommended) {any = "cpu";}"
      return ${$1}
  
      }
// Get integrated recommendations;
  }
  hw_recommender) {any = MockHardwareRecommender();
  integrated_results) { any: any: any = active_learning.integrate_with_hardware_recommender(;
    hardware_recommender: any: any: any = hw_recommender,;
    test_budget: any: any: any = 20,;
    optimize_for: any: any: any = "throughput";"
  );
// Create a batch from the integrated recommendations;
  batch: any: any: any = active_learning.suggest_test_batch(;
    configurations: any: any: any = integrated_results["recommendations"],;"
    batch_size: any: any: any = 10,;
    ensure_diversity: any: any: any = true;
  );
  
  logger.info(`$1`);
  console.log($1);
  console.log($1)}");"
  console.log($1);
  console.log($1).to_dict()}");"
  console.log($1).to_dict()}");"
  
  return batch;

$1($2) {/** Run all test cases. */;
  test_basic_batch_generation();
  test_hardware_constrained_batch();
  test_hardware_availability();
  test_diversity_weighting();
  test_combined_constraints();
  test_integration_with_hardware_recommender()}
  logger.info("All tests completed successfully!");"

$1($2) {
  /** Main function to run tests based on command line arguments. */;
  parser: any: any: any = argparse.ArgumentParser(description="Test the Test Batch Generator functionality");"
  parser.add_argument("--test", choices: any: any: any = ["basic", "hardware", "availability", ;"
                    "diversity", "combined", "integration", "all"],;"
            default: any: any = "all", help: any: any: any = "Test to run");"
  parser.add_argument("--batch-size", type: any: any = int, default: any: any: any = 10,;"
            help: any: any: any = "Batch size for ((test generation") {;"
  parser.add_argument("--verbose", action) { any) {any = "store_true",;"
            help: any: any: any = "Enable verbose output");}"
  args: any: any: any = parser.parse_args();
  
  if ($1) {logging.getLogger().setLevel(logging.DEBUG)}
  if ($1) {
    test_basic_batch_generation();
  elif ($1) {
    test_hardware_constrained_batch();
  elif ($1) {
    test_hardware_availability();
  elif ($1) {
    test_diversity_weighting();
  elif ($1) {
    test_combined_constraints();
  elif ($1) {
    test_integration_with_hardware_recommender();
  elif ($1) {run_all_tests()}
if ($1) {main()}
  }
  };
  };

File: tests/test/predictive_performance/test_batch_generator.py
#!/usr/bin/env python3
"""
Test Batch Generator Testing Script.

This script demonstrates and validates the Test Batch Generator functionality
of the Active Learning System, which is used to create optimized batches of
test configurations for benchmarking.
"""

import os
import sys
import json
import argparse
import logging
import pandas as pd
import numpy as np
from pathlib import Path
from pprint import pprint

# Add parent directory to the Python path to allow importing the module
sys.path.append(str(Path(__file__).parent.parent))

# Import the active learning module directly from the file
import importlib.util
spec = importlib.util.spec_from_file_location(
    "active_learning", 
    str(Path(__file__).parent / "active_learning.py")
)
active_learning_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(active_learning_module)
ActiveLearningSystem = active_learning_module.ActiveLearningSystem

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger("test_batch_generator")

def setup_test_data():
    """Create test data for batch generation."""
    # Create an instance of the active learning system
    active_learning = ActiveLearningSystem()
    
    # Generate test configurations (using the system's built-in function)
    configs = active_learning.recommend_configurations(budget=50)
    
    # Add some metadata for better visualization
    for i, config in enumerate(configs):
        config["id"] = i
        
    return active_learning, configs

def test_basic_batch_generation():
    """Test basic batch generation without special constraints."""
    logger.info("Testing basic batch generation")
    active_learning, configs = setup_test_data()
    
    # Generate a batch with default settings
    batch = active_learning.suggest_test_batch(
        configurations=configs,
        batch_size=10,
        ensure_diversity=True
    )
    
    logger.info(f"Generated batch with {len(batch)} configurations")
    print("\nBasic Batch Generation Results:")
    print(f"Original configurations: {len(configs)}")
    print(f"Batch size: {len(batch)}")
    print(f"Selected configurations (first 3):")
    print(batch[['model_type', 'hardware', 'batch_size', 'expected_information_gain', 'selection_order']].head(3))
    
    # Validate that the batch has the right size
    assert len(batch) <= 10, f"Batch size should be ≤ 10, got {len(batch)}"
    
    # Validate that selection_order column was added
    assert 'selection_order' in batch.columns, "Batch should have selection_order column"
    
    return batch

def test_hardware_constrained_batch():
    """Test batch generation with hardware constraints."""
    logger.info("Testing hardware-constrained batch generation")
    active_learning, configs = setup_test_data()
    
    # Define hardware constraints
    hardware_constraints = {
        'cpu': 2,
        'cuda': 3,
        'openvino': 1,
        'webgpu': 1
    }
    
    # Generate a batch with hardware constraints
    batch = active_learning.suggest_test_batch(
        configurations=configs,
        batch_size=10,
        ensure_diversity=True,
        hardware_constraints=hardware_constraints
    )
    
    logger.info(f"Generated hardware-constrained batch with {len(batch)} configurations")
    print("\nHardware-Constrained Batch Generation Results:")
    print(f"Original configurations: {len(configs)}")
    print(f"Batch size: {len(batch)}")
    
    # Check hardware counts
    hw_counts = batch['hardware'].value_counts().to_dict()
    print(f"Hardware counts: {hw_counts}")
    
    # Validate hardware constraints
    for hw, limit in hardware_constraints.items():
        count = hw_counts.get(hw, 0)
        assert count <= limit, f"Hardware {hw} exceeded constraint: {count} > {limit}"
    
    return batch

def test_hardware_availability():
    """Test batch generation with hardware availability factors."""
    logger.info("Testing hardware availability weighting")
    active_learning, configs = setup_test_data()
    
    # Define hardware availability (probabilities of 0-1)
    hardware_availability = {
        'cpu': 1.0,    # Fully available
        'cuda': 0.8,   # 80% available
        'openvino': 0.5, # 50% available
        'webgpu': 0.2  # 20% available
    }
    
    # Generate a batch with hardware availability weighting
    batch = active_learning.suggest_test_batch(
        configurations=configs,
        batch_size=10,
        ensure_diversity=True,
        hardware_availability=hardware_availability
    )
    
    logger.info(f"Generated availability-weighted batch with {len(batch)} configurations")
    print("\nHardware Availability Batch Generation Results:")
    print(f"Original configurations: {len(configs)}")
    print(f"Batch size: {len(batch)}")
    
    # Check hardware counts
    hw_counts = batch['hardware'].value_counts().to_dict()
    print(f"Hardware counts: {hw_counts}")
    
    # No strict validation here, but we can observe the distribution trends
    
    return batch

def test_diversity_weighting():
    """Test batch generation with different diversity weights."""
    logger.info("Testing diversity weighting impact")
    active_learning, configs = setup_test_data()
    
    results = {}
    
    # Test different diversity weights
    for weight in [0.1, 0.5, 0.9]:
        batch = active_learning.suggest_test_batch(
            configurations=configs,
            batch_size=10,
            ensure_diversity=True,
            diversity_weight=weight
        )
        
        results[weight] = batch
        
        logger.info(f"Generated batch with diversity weight {weight}")
    
    print("\nDiversity Weighting Results:")
    print("Hardware distribution with different diversity weights:")
    
    for weight, batch in results.items():
        hw_counts = batch['hardware'].value_counts().to_dict()
        model_counts = batch['model_type'].value_counts().to_dict()
        print(f"\nDiversity weight {weight}:")
        print(f"Hardware distribution: {hw_counts}")
        print(f"Model type distribution: {model_counts}")
    
    # The higher the diversity weight, the more evenly distributed the configs should be
    
    return results

def test_combined_constraints():
    """Test batch generation with both hardware constraints and availability."""
    logger.info("Testing combined constraints")
    active_learning, configs = setup_test_data()
    
    # Define constraints
    hardware_constraints = {
        'cpu': 3,
        'cuda': 3,
        'openvino': 2,
        'webgpu': 1
    }
    
    hardware_availability = {
        'cpu': 1.0,
        'cuda': 0.7,
        'openvino': 0.5,
        'webgpu': 0.3
    }
    
    # Generate batch with combined constraints
    batch = active_learning.suggest_test_batch(
        configurations=configs,
        batch_size=10,
        ensure_diversity=True,
        hardware_constraints=hardware_constraints,
        hardware_availability=hardware_availability,
        diversity_weight=0.6
    )
    
    logger.info(f"Generated batch with combined constraints: {len(batch)} configurations")
    print("\nCombined Constraints Batch Generation Results:")
    print(f"Original configurations: {len(configs)}")
    print(f"Batch size: {len(batch)}")
    print(f"Hardware counts: {batch['hardware'].value_counts().to_dict()}")
    print(f"Model type counts: {batch['model_type'].value_counts().to_dict()}")
    
    # Validate hardware constraints
    hw_counts = batch['hardware'].value_counts().to_dict()
    for hw, limit in hardware_constraints.items():
        count = hw_counts.get(hw, 0)
        assert count <= limit, f"Hardware {hw} exceeded constraint: {count} > {limit}"
    
    return batch

def test_integration_with_hardware_recommender():
    """
    Test integration between batch generation and hardware recommender.
    
    This is a simulated test since we don't have actual hardware recommender available.
    """
    logger.info("Testing batch generation integration with hardware recommender")
    active_learning, _ = setup_test_data()
    
    # Create mock hardware recommender
    class MockHardwareRecommender:
        def recommend_hardware(self, **kwargs):
            model_type = kwargs.get("model_type", "")
            
            # Simple logic to simulate hardware recommendations
            if model_type == "text_embedding":
                recommended = "cuda"
            elif model_type == "vision":
                recommended = "webgpu"
            elif model_type == "audio":
                recommended = "openvino"
            else:
                recommended = "cpu"
                
            return {
                "recommended_hardware": recommended,
                "throughput_score": 0.8,
                "estimated_improvement": 0.3,
                "alternatives": ["cpu"]
            }
    
    # Get integrated recommendations
    hw_recommender = MockHardwareRecommender()
    integrated_results = active_learning.integrate_with_hardware_recommender(
        hardware_recommender=hw_recommender,
        test_budget=20,
        optimize_for="throughput"
    )
    
    # Create a batch from the integrated recommendations
    batch = active_learning.suggest_test_batch(
        configurations=integrated_results["recommendations"],
        batch_size=10,
        ensure_diversity=True
    )
    
    logger.info(f"Generated integrated batch with {len(batch)} configurations")
    print("\nIntegrated Batch Generation Results:")
    print(f"Hardware recommendations: {len(integrated_results['recommendations'])}")
    print(f"Batch size: {len(batch)}")
    print(f"Hardware counts: {batch['hardware'].value_counts().to_dict()}")
    print(f"Model type counts: {batch['model_type'].value_counts().to_dict()}")
    
    return batch

def run_all_tests():
    """Run all test cases."""
    test_basic_batch_generation()
    test_hardware_constrained_batch()
    test_hardware_availability()
    test_diversity_weighting()
    test_combined_constraints()
    test_integration_with_hardware_recommender()
    
    logger.info("All tests completed successfully!")

def main():
    """Main function to run tests based on command line arguments."""
    parser = argparse.ArgumentParser(description="Test the Test Batch Generator functionality")
    parser.add_argument("--test", choices=["basic", "hardware", "availability", 
                                          "diversity", "combined", "integration", "all"],
                        default="all", help="Test to run")
    parser.add_argument("--batch-size", type=int, default=10,
                        help="Batch size for test generation")
    parser.add_argument("--verbose", action="store_true",
                        help="Enable verbose output")
    
    args = parser.parse_args()
    
    if args.verbose:
        logging.getLogger().setLevel(logging.DEBUG)
    
    if args.test == "basic":
        test_basic_batch_generation()
    elif args.test == "hardware":
        test_hardware_constrained_batch()
    elif args.test == "availability":
        test_hardware_availability()
    elif args.test == "diversity":
        test_diversity_weighting()
    elif args.test == "combined":
        test_combined_constraints()
    elif args.test == "integration":
        test_integration_with_hardware_recommender()
    elif args.test == "all":
        run_all_tests()

if __name__ == "__main__":
    main()